src/hotspot/share/opto/compile.cpp

```diff
@@ -1350,7 +1350,7 @@ const TypePtr *Compile::flatten_alias_type( const TypePtr *tj ) const {
       } else if( offset == arrayOopDesc::length_offset_in_bytes() ) {
         // range is OK as-is.
         tj = ta = TypeAryPtr::RANGE;
-      } else if( offset == oopDesc::klass_offset_in_bytes() ) {
+      } else if( offset == TypeOopPtr::klass_offset_in_bytes() ) {
```

Comment:
Not sure why this is here.

Reply:
The reason I did it this way is so that I can assert(!UCOH) in oopDesc::klass_offset_in_bytes(). The offset is used in C2 only as a sort of marker (to distinguish memory slice for the klass-word from other memory slices). It is not actually used as offset to load the Klass* - we load the whole mark-word and shift/mask/expand the Klass* from that, including the check for monitor.
In the long run, as soon as the monitor-mapping in the header goes away, we could indeed use the oopDesc::klass_offset_in_bytes() everywhere - unless when we squeeze the narrowKlass in the header at a non-byte boundery, which is bound to happen once we go to 4-byte headers.

Reply:
My concern is that it deviates from what we have in "mainline" version: https://builds.shipilev.net/patch-openjdk-lilliput/src/hotspot/share/opto/compile.cpp.sdiff.html -- which means we would always need to look for these discrepancies when backporting.

How does `openjdk/lilliput` achieve this?

Reply:
> My concern is that it deviates from what we have in "mainline" version: https://builds.shipilev.net/patch-openjdk-lilliput/src/hotspot/share/opto/compile.cpp.sdiff.html -- which means we would always need to look for these discrepancies when backporting.
> 
> How does `openjdk/lilliput` achieve this?

It doesn't assert !UCOH in oopDesc::klass_offset_in_bytes(). Best approaches will be to either remove the assert and keep the 21 (and jdk upstream) stuff simpler, at the cost of loosing the assert, or bring over the C2 changes.

Reply:
Lilliput JDK 17 does not do this either, it seems: https://builds.shipilev.net/patch-openjdk-lilliput-jdk17/src/hotspot/share/opto/compile.cpp.sdiff.html

So, I suggest we keep 21u close to both mainline and 17u by not asserting too, and using the usual `oopDesc::`?